File: tf.nim
import os
import sequtils
import streams
import strutils
import tables

import docopt

let doc = """
Topfew.

Usage:
  tf [--fields=<fields>] [--few=<few>] <filename>

Options:
  --fields=<fields>   Which fields to select [default: 1]
  --few=<few>         How many results to display [default: 10]
"""


proc findKey(s: string, start: int, occ: int, actualOcc: int): (int, string) =
    let found = find(s, ' ', start)
    if found < 1:
        return (-1, "")
    if actualOcc + 1 == occ:
        return (found, s[start..found])
    return findKey(s, found + 1, occ, actualOcc + 1)


when isMainModule:
    let args = docopt(doc, version = "Topfew 0.1")

    let fieldsArg = split($args["--fields"], ',')
    var fields: seq[int] = @[]
    for x in fieldsArg:
        add(fields, parseInt(x))
    let few = parseInt($args["--few"])
    let filename = $args["<filename>"]

    if not fileExists(filename):
        echo "Page " & filename & " does not exist"
        quit 1

    var lines: seq[seq[string]] = @[]
    var strm = newFileStream(filename, fmRead)
    var line = ""
    var keys = initCountTable[string]()
    while strm.readLine(line):
        var key = ""
        var tok = ""
        var count = 0
        var pos = 0
        var lastField = 0
        for field in fields:
            (pos, tok) = findKey(line, pos, field, lastField)
            lastField = field
            if pos < 1:
                raise newException(ValueError, "Field " & field.`$` & " not found")
            key = key & tok
        inc(keys, key)
    sort(keys)
    var count = 0
    for k,v in pairs(keys):
        echo v.`$` & " " & k
        count = count + 1
        if count >= few:
            break

File: tf.py
#! /usr/bin/env python3

"""
Topfew.

Usage:
  tf [--fields=<fields>] [--few=<few>] <filename>

Options:
  --fields=<fields>   Which fields to select [default: 1]
  --few=<few>         How many results to display [default: 10]
"""

from docopt import docopt


args = docopt(__doc__, version="0.1")

fields = [int(x) for x in args["--fields"].split(',')]
few = int(args["--few"])
filename = args["<filename>"]

keys = {}
with open(filename) as f:
    for line in f.readlines():
        key = ""
        tokens = line.split(' ')
        for field in fields:
            key = key + tokens[field-1]
        if key not in keys:
            keys[key] = 1
        else:
            keys[key] = keys[key] +1

count = 0
for k,v in {k: v for k, v in sorted(keys.items(), key=lambda item: item[1], reverse=True)}.items():
    print(v, k)
    count += 1
    if count >= few:
        break
